Thoroughly nospecialize all functions; add no-alloc, no-specialize test.

NHDaly opened the pull request:
Make sure that _all_ functions in ExceptionUnwrapping are marked nospecialize. We don't want to pay the wasted compilation time at runtime, since these are all going to be called in _exceptional_ cases, certainly not in a hot loop, and because we've seen crashes caused by a stackoverflow in type inference while attempting to specialize the code to handle a StackOverflowException! 😅

This time, we add a unit test to ensure that that these functions do not allocate and do not incur new compilation when called with novel arguments.

This is a follow-up to #6 and #15.

kpamnany commented:
FYI: some type piracy? https://github.com/RelationalAI/raicode/blob/master/packages/Arroyo/src/stack_frames.jl#L248

NHDaly committed:
Thoroughly nospecialize all functions; add no-alloc, no-specialize test.

Make sure that _all_ functions in ExceptionUnwrapping are marked
nospecialize. We don't want to pay the wasted compilation time at
runtime, since these are all going to be called in _exceptional_ cases,
certainly not in a hot loop, and because we've seen crashes caused by a
stackoverflow in type inference while attempting to specialize the code
to handle a StackOverflowException! 😅

This time, we add a unit test to ensure that that these functions do not
allocate and do not incur new compilation when called with novel
arguments.

Drvi commented on src/exception_summary.jl:
```diff
@@ -110,7 +112,10 @@ function _summarize_exception(io::IO, e::CompositeException, stack; prefix = not
 # This is the overload that prints the actual exception that occurred.
-function _summarize_exception(io::IO, exc, stack; prefix = nothing)
+function _summarize_exception(io::IO, @nospecialize(exc), stack; prefix = nothing)
+    @show exc
+    @show is_wrapped_exception(exc)
+    global EXC = exc
```
I think this is a leftover from debugging
